Userland/Libraries/LibAudio/ConnectionFromClient.cpp

```diff
@@ -1,48 +1,123 @@
-static timespec g_enqueue_wait_time { 0, 10'000'000 };
-
 ConnectionFromClient::ConnectionFromClient(NonnullOwnPtr<Core::Stream::LocalSocket> socket)
     : IPC::ConnectionToServer<AudioClientEndpoint, AudioServerEndpoint>(*this, move(socket))
+    , m_buffer(make<AudioQueue>(MUST(AudioQueue::try_create())))
+    , m_user_queue(make<UserSampleQueue>())
+    , m_background_audio_enqueuer(Threading::Thread::construct([this]() {
+        // All the background thread does is run an event loop
+        Core::EventLoop enqueuer_loop;
+        m_enqueuer_loop = &enqueuer_loop;
+        enqueuer_loop.exec();
+        m_enqueuer_loop = nullptr;
+        return (intptr_t) nullptr;
+    }))
+{
+    m_background_audio_enqueuer->start();
+    set_buffer(*m_buffer);
+}
+
+ConnectionFromClient::~ConnectionFromClient()
+{
+    die();
+}
+
+void ConnectionFromClient::die()
+{
+    // We're sometimes getting here after the other thread has already exited and its event loop does no longer exist.
+    if (m_enqueuer_loop != nullptr) {
+        m_enqueuer_loop->wake();
+        m_enqueuer_loop->quit(0);
+    }
+    (void)m_background_audio_enqueuer->join();
+}
+
+ErrorOr<void> ConnectionFromClient::async_enqueue(FixedArray<Sample>&& samples)
 {
+    update_good_sleep_time();
+    m_user_queue->append(move(samples));
+    // Wake the background thread to make sure it starts enqueuing audio.
+    if (!m_audio_enqueuer_active.load())
+        m_enqueuer_loop->post_event(*this, make<Core::CustomEvent>(0), Core::EventLoop::ShouldWake::Yes);
+    async_start_playback();
+
+    return {};
+}
+
+void ConnectionFromClient::update_good_sleep_time()
+{
+    auto sample_rate = static_cast<double>(get_sample_rate());
+    auto buffer_play_time_ns = 1'000'000'000.0 / (sample_rate / static_cast<double>(AUDIO_BUFFER_SIZE));
+    // A factor of 1 should be good for now.
+    m_good_sleep_time = static_cast<unsigned>(buffer_play_time_ns);
 }
 
-void ConnectionFromClient::enqueue(LegacyBuffer const& buffer)
+// Non-realtime audio writing loop
+void ConnectionFromClient::custom_event(Core::CustomEvent&)
 {
-    for (;;) {
-        auto success = enqueue_buffer(buffer.anonymous_buffer(), buffer.id(), buffer.sample_count());
-        if (success)
+    while (true) {
+        m_audio_enqueuer_active.store(true);
+        Array<Sample, AUDIO_BUFFER_SIZE> next_chunk;
+
+        if (m_user_queue->is_empty()) {
+            dbgln("Reached end of provided audio data, going to sleep");
             break;
-        nanosleep(&g_enqueue_wait_time, nullptr);
+        }
+
+        auto available_samples = min(AUDIO_BUFFER_SIZE, m_user_queue->size());
+        for (size_t i = 0; i < available_samples; ++i) {
+            next_chunk[i] = (*m_user_queue)[i];
+        }
+
+        m_user_queue->slice(available_samples);
+
+        auto result = m_buffer->try_blocking_enqueue(next_chunk, [this]() {
+            timespec const sleep_time {
+                .tv_sec = m_good_sleep_time / 1'000'000'000,
```

Comment:
`AK::Time` should really be used to do time conversions. 
Also, why not just store the time you want to sleep as a `timespec` so you don't need to convert on every sleep?